File: 프로그래머스/2/42584. 주식가격/주식가격.cpp
#include<iostream>
#include<stack>
#include<vector>
#include<queue>

using namespace std;

vector<int>solution(vector<int>prices) {
	vector<int> answer;
	queue<int>q;
	// 주식 가격들을 큐에 삽입
	for (int i = 0; i < prices.size(); i++) {
		q.push(prices[i]);
	}
	//큐에 남아있는 주식 가격을 차례대로 처리
	while (q.size()) {
		for (int i = 0; i < prices.size(); i++) {
			int tmp = q.front();
			int count = 0;// 떨어지지 않은 기간 누적 카운트
			q.pop();
			if (i + 1 < prices.size()) {
				for (int j = i + 1; j < prices.size(); j++) {
					count++;
					if (tmp > prices[j])
						break;//기준 값이 더 크다면 가격이 떨어진 것이므로 루프 종료
					
				}
			}
			answer.push_back(count);//카운트를 벡터에 저장
		}
	}
	return answer;//결과 벡터를 리턴
}
int main() {
	vector<int> ex = { 1,2,3,2,3 };
	vector<int>result = solution(ex);
	for (int i = 0; i < result.size(); i++) {
		cout << result[i] << endl;
	}
	return 0;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;
class Solution {
    public int[] solution(int[] prices) {
        int n = prices.length;
        int[] answer = new int[n];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            // 현재 가격이 스택 top보다 낮아졌다면 answer에 계산
            while (!stack.isEmpty() && prices[i] < prices[stack.peek()]) {
                int top = stack.pop();
                answer[top] = i - top;
            }
            stack.push(i);
        }

        // 끝까지 가격이 떨어지지 않은 인덱스 처리
        while (!stack.isEmpty()) {
            int top = stack.pop();
            answer[top] = n - 1 - top;
        }

        return answer;
    }
}